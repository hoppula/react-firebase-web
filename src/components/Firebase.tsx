import * as firebase from "firebase"
import * as PropTypes from "prop-types"
import * as React from "react"

export interface FirebaseProps {
  readonly apiKey: string
  readonly projectId: string
  readonly databaseURL?: string
}

export class Firebase extends React.Component<FirebaseProps, {}> {
  static propTypes = {
    apiKey: PropTypes.string.isRequired,
    databaseURL: PropTypes.string,
    projectId: PropTypes.string.isRequired
  }

  static childContextTypes = {
    firebase: PropTypes.object
  }

  firebase: firebase.app.App

  constructor(props: FirebaseProps) {
    super(props)
    const { apiKey, projectId, databaseURL } = props
    this.firebase = firebase.initializeApp({
      apiKey,
      authDomain: `${projectId}.firebaseapp.com`,
      databaseURL: databaseURL
        ? databaseURL
        : `https://${projectId}.firebaseio.com`,
      storageBucket: `${projectId}.appspot.com`
    })
  }

  getChildContext() {
    return {
      firebase: this.firebase
    }
  }

  render() {
    return React.Children.only(this.props.children)
  }
}

export default Firebase
