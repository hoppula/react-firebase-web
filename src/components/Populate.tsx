import * as firebase from "firebase"
import * as PropTypes from "prop-types"
import * as React from "react"
import findIndex = require("lodash.findindex")
import { FirebaseCallback, FirebaseContext } from "../types"

export type KeyValue = { key: string; value: any }

export interface PopulateProps {
  readonly from: { [key: string]: boolean }
  readonly children: (value: any) => React.ReactElement<any>
  readonly with: (key: string) => string
}

export interface PopulateState {
  value: KeyValue[]
}

export class Populate extends React.Component<PopulateProps, PopulateState> {
  static propTypes = {
    from: PropTypes.any,
    children: PropTypes.func.isRequired,
    with: PropTypes.func.isRequired
  }

  static contextTypes = {
    firebase: PropTypes.object.isRequired
  }

  context: FirebaseContext
  listeners: {
    ref: firebase.database.Reference
    callback: FirebaseCallback
  }[]

  state = {
    value: [] as KeyValue[]
  }

  componentDidMount() {
    const { from, with: populateWith } = this.props
    this.addListeners(from, populateWith)
  }

  addListeners(
    from: { [key: string]: boolean },
    populateWith: (key: string) => string
  ) {
    const keys = Object.keys(from || {})
    const refs = keys.map(key => {
      return this.context.firebase.database().ref(populateWith(key))
    })
    this.listeners = refs.map(ref => {
      return {
        ref,
        callback: ref.on("value", snapshot => {
          this.setState(state => {
            const nextValue = [...state.value]
            const replaceIndex = findIndex(keys, key => key === snapshot.key)
            if (replaceIndex >= 0) {
              nextValue[replaceIndex] = {
                key: snapshot.key,
                value: snapshot.val()
              }
            }
            return { value: nextValue }
          })
        })
      }
    })
  }

  removeListeners() {
    this.listeners.forEach(listener => {
      listener.ref.off("value", listener.callback)
    })
  }

  componentDidUpdate(prevProps: PopulateProps) {
    const { from, with: populateWith } = this.props
    const { from: prevFrom, with: prevWith } = prevProps
    if (from !== prevFrom || populateWith !== prevWith) {
      this.removeListeners()
      this.addListeners(from, populateWith)
    }
  }

  componentWillUnmount() {
    this.removeListeners()
  }

  render() {
    return React.Children.only(this.props.children(this.state.value))
  }
}

export default Populate
