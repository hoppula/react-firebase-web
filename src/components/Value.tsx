import * as firebase from "firebase"
import * as PropTypes from "prop-types"
import * as React from "react"

import { FirebaseCallback, FirebaseContext } from "../types"

export interface ValueProps {
  readonly once?: boolean
  readonly path: string
  readonly children: (value: any) => React.ReactElement<any>
}

export interface ValueState {
  value: any
}

export class Value extends React.Component<ValueProps, ValueState> {
  static propTypes = {
    once: PropTypes.bool,
    path: PropTypes.string.isRequired,
    children: PropTypes.func.isRequired
  }

  static contextTypes = {
    firebase: PropTypes.object.isRequired
  }

  static defaultProps = {
    once: false
  }

  context: FirebaseContext
  private ref: firebase.database.Reference
  private listeners: {
    value: FirebaseCallback
  }

  constructor(props: ValueProps) {
    super(props)
    this.state = {
      value: {}
    }
  }

  componentDidMount() {
    const { path, once } = this.props
    this.ref = this.createRef(path)
    this.listeners = {
      value: null
    }
    this.addListeners(this.ref, once)
  }

  createRef(path: string) {
    return this.context.firebase.database().ref(path)
  }

  addListeners(ref: firebase.database.Reference, once: boolean) {
    const listeners = this.listeners

    // listen only once if once prop has been passed
    if (once) {
      return ref.once("value", snapshot => {
        this.setState({
          value: snapshot.val()
        })
      })
    }

    // use live bindings as default
    listeners.value = ref.on("value", snapshot => {
      this.setState(() => {
        return { value: snapshot.val() }
      })
    })
  }

  removeListeners() {
    Object.keys(this.listeners).forEach((eventType: "value") => {
      this.ref.off(eventType, this.listeners[eventType])
    })
    this.listeners = {
      value: null
    }
  }

  componentDidUpdate(prevProps: ValueProps) {
    const { path, once } = this.props
    const { path: prevPath, once: prevOnce } = prevProps
    if (path !== prevPath || once !== prevOnce) {
      this.removeListeners()
      this.ref = this.createRef(path)
      this.addListeners(this.ref, once)
    }
  }

  componentWillUnmount() {
    this.removeListeners()
  }

  render() {
    return React.Children.only(this.props.children(this.state.value))
  }
}

export default Value
