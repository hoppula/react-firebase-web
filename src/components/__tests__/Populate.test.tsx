import React from "react"
import renderer from "react-test-renderer"
import { Firebase } from "../Firebase"
import Populate from "../Populate"
import Value from "../Value"

import { initializeMockSDK } from "./setup"
const mocksdk = initializeMockSDK()

jest.mock("firebase/app", () => {
  return {
    initializeApp: (...args: any[]) => mocksdk.initializeApp(...args)
  }
})

describe("Populate", () => {
  test("populates related lists properly", () => {
    let component = renderer.create(
      <Firebase apiKey="firebase-test" projectId="firebase-test">
        <Value
          path="user"
          children={user => {
            const bookmarked = user ? user.bookmarkedArticles : {}
            return (
              <Populate
                from={bookmarked}
                with={articledId => `articles/${articledId}`}
                children={list => (
                  <ul>
                    {list.map(
                      ({
                        key,
                        value: article
                      }: {
                        key: string
                        value: any
                      }) => (
                        <li key={key}>{article.title}</li>
                      )
                    )}
                  </ul>
                )}
              />
            )
          }}
        />
      </Firebase>
    )

    const dbRef = component.getInstance().firebase.database()
    dbRef.ref("user").set({ username: "Test user", bookmarkedArticles: [] })
    dbRef.ref("user").flush()

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot("list should be empty")

    const articlesRef = dbRef.ref("articles")
    articlesRef.push({ title: "First article" })
    articlesRef.push({ title: "Second article" })
    articlesRef.flush()

    const articleIds = Object.keys(articlesRef.getData())

    articleIds.forEach(articleId => {
      dbRef.ref(`user/bookmarkedArticles/${articleId}`).set(true)
    })
    dbRef.ref("user/bookmarkedArticles").flush()

    tree = component.toJSON()
    expect(tree).toMatchSnapshot("all bookmarked articles should be rendered")

    const third = articlesRef.push({ title: "Third article" })
    articlesRef.flush()
    dbRef.ref(`user/bookmarkedArticles/${third.key}`).set(true)
    dbRef.ref("user/bookmarkedArticles").flush()

    tree = component.toJSON()
    expect(tree).toMatchSnapshot(
      "third article should have been added to bookmarks"
    )

    third.update({ title: "Third article with new title" })
    articlesRef.flush()

    tree = component.toJSON()
    expect(tree).toMatchSnapshot("third article should have new title")
  })
})
