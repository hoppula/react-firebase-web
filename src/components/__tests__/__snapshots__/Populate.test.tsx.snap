// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Populate populates related lists properly: all bookmarked articles should be rendered 1`] = `
<ul>
  <li>
    First article
  </li>
  <li>
    Second article
  </li>
</ul>
`;

exports[`Populate populates related lists properly: list should be empty 1`] = `<ul />`;

exports[`Populate populates related lists properly: third article should have been added to bookmarks 1`] = `
<ul>
  <li>
    First article
  </li>
  <li>
    Second article
  </li>
  <li>
    Third article
  </li>
</ul>
`;

exports[`Populate populates related lists properly: third article should have new title 1`] = `
<ul>
  <li>
    First article
  </li>
  <li>
    Second article
  </li>
  <li>
    Third article with new title
  </li>
</ul>
`;
